model/src/main/kotlin/FileList.kt

```diff
@@ -64,4 +64,12 @@ data class FileList(
             }
         }
     }
+
+    operator fun plus(other: FileList): FileList {
+        require(provenance == other.provenance) {
+            "Cannot merge FileLists with different provenance: $provenance != ${other.provenance}."
+        }
+
+        return copy(files = files + other.files)
```

Comment:
I'm a little bit concerned that with the implementation it's to easy to miss to also add new properties here if we ever add new properties to the class.

So how about instead implementing this like (omitting the custom message for brevity)

```kotlin
operator fun plus(other: FileList) =
    FileList(
        provenance = provenance.also { require(it == other.provenance) },
        files = files + other.files
    )
```

which would ensure that no property can be missed?